cpp/src/strings/copying/copy_range.cu

```diff
@@ -0,0 +1,137 @@
+    } else {
+      return target_has_nulls && d_target.is_null_nocheck(idx)
+               ? 0
+               : d_target.element<string_view>(idx).size_bytes();
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> copy_range(strings_column_view const& source,
+                                   strings_column_view const& target,
+                                   size_type source_begin,
+                                   size_type source_end,
+                                   size_type target_begin,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  auto target_end = target_begin + (source_end - source_begin);
+  CUDF_EXPECTS(
+    (target_begin >= 0) && (target_begin < target.size()) && (target_end <= target.size()),
+    "Range is out of bounds.");
+
+  if (target_end == target_begin) { return std::make_unique<column>(target.parent(), stream, mr); }
+  auto source_device_view = column_device_view::create(source.parent(), stream);
+  auto d_source           = *source_device_view;
+  auto target_device_view = column_device_view::create(target.parent(), stream);
+  auto d_target           = *target_device_view;
+
+  // create null mask
+  auto [null_mask, null_count] = cudf::detail::valid_if(
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(target.size()),
+    [d_source, d_target, source_begin, target_begin, target_end] __device__(size_type idx) {
+      return (idx >= target_begin && idx < target_end)
+               ? d_source.is_valid(source_begin + (idx - target_begin))
+               : d_target.is_valid(idx);
+    },
+    stream,
+    mr);
+
+  auto [check_source, check_target] = [target, null_count = null_count] {
+    // check validities for both source & target
+    if (target.has_nulls()) { return std::make_pair(true, true); }
+    // check validities for source only
+    if (null_count > 0) { return std::make_pair(true, false); }
+    // no need to check validities
+    return std::make_pair(false, false);
+  }();
+
+  // create offsets
+  auto sizes_begin = cudf::detail::make_counting_transform_iterator(
+    0,
+    compute_element_size{
+      d_source, d_target, source_begin, target_begin, target_end, check_source, check_target});
+  auto [offsets_column, chars_bytes] = cudf::strings::detail::make_offsets_child_column(
+    sizes_begin, sizes_begin + target.size(), stream, mr);
+  auto d_offsets = cudf::detail::offsetalator_factory::make_input_iterator(offsets_column->view());
+
+  // create chars
+  auto chars_data = rmm::device_uvector<char>(chars_bytes, stream, mr);
+  auto d_chars    = chars_data.data();
+  thrust::for_each(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(target.size()),
+    [d_source, d_target, source_begin, target_begin, target_end, d_offsets, d_chars] __device__(
```

Comment:
maybe cuda::proclaim_return_type, not sure if it's needed for void

Reply:
Not needed for void or certain thrust calls that take a predicate (function returning bool).

Reply:
You’ll know if it is needed because compilation will fail.

Reply:
So I only usually put it in if the compile fails without it. Hopefully that is ok.

Reply:
Thank god, it's not another thing to check for in code review. Thanks for the clarification.